### PR TITLE
Improve pubsub integration test

### LINKED ISSUE #1610: Google.Cloud.PubSub.V1.IntegrationTests BulkMessaging test hangs on deadlock
Jon, Chris: Not sure which of you is best to talk to on this.  This is showing up on the Windows CI.  Just using this as a place holder so we don't lose track.

I have had two tests run that hang for about 30 min.

### PR DESCRIPTION
**Pull request by chrisdunelm**
Handle duplicate messages and nacks properly, hopefully make it less flakey

Fix #1610

### CODE REVIEW

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/SimplePubSubTest.cs [hunk 1]
```diff
@@ -41,13 +42,18 @@ public SimplePubSubTest(PubsubFixture fixture)
 
         private readonly PubsubFixture _fixture;
 
-        private async Task RunBulkMessaging(int messageCount, int minMessagesize, int maxMessageSize, int maxMessagesInFlight, int initialNackCount, TimeSpan? timeouts = null)
+        private async Task RunBulkMessaging(
+            int messageCount, int minMessageSize, int maxMessageSize, int maxMessagesInFlight, int initialNackCount,
+            TimeSpan? timeouts = null, int? cancelAfterRecvCount = null)
         {
+            // Force messages to be at least 4 byte long, so an int ID can be used.
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/SimplePubSubTest.cs]: Nit: byte => bytes

Comment (jskeet) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/SimplePubSubTest.cs]: (And don't we control the arguments? Can't we just validate that minMessageSize is at least 4?)

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/SimplePubSubTest.cs [hunk 2]
```diff
@@ -73,29 +79,50 @@ public SimplePubSubTest(PubsubFixture fixture)
...
+                int id = BitConverter.ToInt32(msg.Data.ToArray(), 0);
+                lock (nackedIds)
                 {
-                    return Task.FromResult(SimpleSubscriber.Reply.Nack);
+                    if (nackedIds.Count < initialNackCount)
+                    {
+                        if (nackedIds.Add(id))
+                        {
+                            // This ID not already nacked
+                            var localRecvCount = Interlocked.Increment(ref recvCount);
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/SimplePubSubTest.cs]: What if this is cancelAfterRecvCount? Or is that always greater than initialNackCount? (Is that checked?)

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/SimplePubSubTest.cs [hunk 3]
```diff
@@ -73,29 +79,50 @@ public SimplePubSubTest(PubsubFixture fixture)
...
+                            return Task.FromResult(SimpleSubscriber.Reply.Nack);
+                        }
+                    }
+                }
+                bool wasAdded = recvedIds.TryAdd(id, false);
+                if (wasAdded)
+                {
+                    var localRecvCount = Interlocked.Increment(ref recvCount);
+                    Interlocked.Add(ref recvSum, msg.Data.Sum(x => (long)x));
+                    if (localRecvCount == cancelAfterRecvCount || localRecvCount >= messageCount + initialNackCount)
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/SimplePubSubTest.cs]: Why would it be greater than messageCount + initialNackCount?

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/SimplePubSubTest.cs (hunk 2 above)
Comment (chrisdunelm) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/SimplePubSubTest.cs]: cancelAfterRecvCount only applies to non-nacked (acked) messages, although I agree that's not very clear.
(Also got rid of pointless allocation)

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/SimplePubSubTest.cs (hunk 3 above)
Comment (chrisdunelm) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/SimplePubSubTest.cs]: It shouldn't be, but this is just in case...

**Pull request merged**